waiter/src/waiter/scheduler/kubernetes.clj

```diff
@@ -0,0 +1,605 @@
+      (catch Throwable e
+        (log/warn "Kubernetes ReplicaSet conflict while deleting"
+                  {:service-id service-id})
+        {:result :error
+         :message "Internal error while deleting service"})))
+
+  (scale-app [this service-id scale-to-instances]
+    (ss/try+
+      (if-let [service (service-id->service this service-id)]
+        (do
+          (scale-service-up-to this service scale-to-instances)
+          {:success true
+           :status 200
+           :result :scaled
+           :message (str "Scaled to " scale-to-instances)})
+        (do
+          (log/error "Cannot scale missing service" service-id)
+          {:success false
+           :status 404
+           :result :no-such-service-exists
+           :message "Failed to scale missing service"}))
+      (catch [:status 409] _
+        {:success false
+         :status 409
+         :result :conflict
+         :message "Scaling failed due to repeated patch conflicts"})
+      (catch Throwable e
+        (log/error e "Error while scaling waiter service" service-id)
+        {:success false
+         :status 500
+         :result :failed
+         :message "Error while scaling waiter service"})))
+
+  (retrieve-directory-content [_ service-id instance-id _ relative-directory]
+    ;; TODO (#357) - get access to the working directory contents in the pod
+    [])
+
+  (service-id->state [_ service-id]
+    {:failed-instances (vals (get @service-id->failed-instances-transient-store service-id))
+     :killed-instances (scheduler/service-id->killed-instances service-id)})
+
+  (state [_]
+    {:service-id->failed-instances @service-id->failed-instances-transient-store}))
+
+(defn- start-auth-renewer
+  "Initialize the k8s-api-auth-str atom,
+   and optionally start a chime to periodically referesh the value."
+  [{:keys [refresh-delay-minutes refresh-fn]}]
+  (let [refresh (-> refresh-fn utils/resolve-symbol deref)
+        auth-update-fn (fn auth-update []
+                         (if-let [auth-str' (refresh)]
+                           (reset! k8s-api-auth-str auth-str')))]
+    (auth-update-fn)
+    (when [refresh-delay-minutes]
+      (assert (utils/pos-int? refresh-delay-minutes))
+      (-> refresh-delay-minutes
+          t/minutes
+          (du/start-timer-task auth-update-fn)))))
+
+(defn kubernetes-scheduler
+  "Returns a new KubernetesScheduler with the provided configuration. Validates the
+  configuration against kubernetes-scheduler-schema and throws if it's not valid."
+  [{:keys [authentication http-options max-conflict-retries max-name-length
+           pod-base-port rs-spec-file-path service-id->service-description-fn url]}]
+  {:pre [(utils/pos-int? (:socket-timeout http-options))
+         (utils/pos-int? (:conn-timeout http-options))
+         (utils/non-neg-int? max-conflict-retries)
+         (utils/pos-int? max-name-length)
+         (integer? pod-base-port)
+         (<= 1024 pod-base-port 49151)  ; registered port range
```

Comment:
Can we add some extra documentation somewhere on why this is the registered port range, e.g. in the `config-full.edn`?

Reply:
If I remember correctly, Java throws if you try to bind to a port above 49151. Ports below 1024 require root, and are more likely to conflict with other system services.

If you agree that those are good reasons to restrict the port range, then I can explain that in config-full.edn. If you thin I'm being overly restrictive here, I can loosen the preconditions.

Reply:
We can follow the intersection of what Mesos agents allow as their port ranges and the ranges that work with Kubernetes.

Reply:
It looks like Mesos lets you specify custom ranges of ports that a slave can expose to containers as resources, but by default uses a small subrange of the "registered" ports: 31000 to 32000.<sup>[link](https://github.com/apache/mesos/blob/0f0909c89578dbe9ea23d5fa4d980712789a2c32/src/slave/constants.hpp#L108-L109)

Reply:
We can set it to 31000 by default.

Should we allow users to configure in any valid positive integer (1 - 65535)?

Reply:
We should probably do 65526 as the upper bound (since we allow using up to 10 ports), but both the 31000 default and the less-restrictive range sound reasonable.

Reply:
Done.